src/itensor.jl

```diff
@@ -1892,8 +1892,13 @@ diag(T::ITensor) = diag(tensor(T))
 
 mul!(C::ITensor, A::ITensor, B::ITensor, args...)::ITensor = contract!(C, A, B, args...)
 
-dot(A::ITensor, B::ITensor) = (dag(A) * B)[]
+## TODO this operation does not work with GPU
+dot(A::ITensor, B::ITensor) = _dot(dag(A), B)
 
+function _dot(A::ITensor, B::ITensor)
+  B = permute(B, inds(A))
+  return dot(tensor(A), tensor(B))
+end
```

Comment:
Is the issue the operation `[]`? If so, seems like in `NDTensors` we could write a specialized version of `getindex(A::Tensor)` where `A` is a tensor backed by a `CuArray` (say by first converting to CPU with `NDTensors.cpu`).

Reply:
I would only propose overloading an empty `getindex` call on a scalar `Tensor`/`ITensor` (i.e. `getindex(T)`, not `getindex(T, I...)`, in which case there is only one element anyway, and also only if that is necessary to make this code work generically without modifying code in ITensors.jl.